File: cpp/PE023.cpp
#include <iostream>
using namespace std;

bool a[28123+1];
int b[28123+1];
int sumOfAllProperDivisors(int n) {
    int result = 0;
    int i;
    for(i = 1; i < n; i++)
        if(n%i == 0)
            result += i;
    return result;
}

bool isAbundant(int n) {
    if(sumOfAllProperDivisors(n) > n)
        return true;
    return false;
}

int main() {
    int i, j = 0, k, result = 0;
    for(i = 1; i <= 28123; i++)
        a[i] = false;
    for(i = 1; i <= 28123; i++){
        if(isAbundant(i))
            b[j++] = i;
    }
    for(i = 0; i < j; i++)
        for(k = i; k < j; k++){
            if(b[i] + b[k] <= 28123)
                a[b[i]+b[k]] = true;
        }
    for(i = 1; i <= 28123; i++){
        if(a[i] == false)
            result += i;
    }
    cout << result << endl;
    return 0;
}

File: java/PE023.java
public class PE023 {
    public static void main(String[] args)
    {
        int sum = 0;
        int j = 0;
        boolean[] a = new boolean[28123+1];
        int[] b = new int[28123+1];
        for(int i = 1; i <= 28123; i++) {
            a[i] = false;
            if(isAbundant(i)){
                b[j++] = i;
            }
        }
        for(int i = 0; i < j; i++) {
            for(int k = 0; k <= i; k++) {
                if(b[k] + b[i] <= 28123)
                    a[b[k]+b[i]] = true;
            }
        }
        for(int i = 0; i < 28123; i++) {
            if(a[i] == false)
                sum += i;
        }
        System.out.println(sum);
    }
    private static boolean isAbundant(int a)
    {
        int sum = 0;
        for(int i = 1; i < a; i++) {
            if(a%i == 0)
                sum += i;
        }
        if(sum > a)
            return true;
        else
            return false;
    }
}
